cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeTable.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright © 2014-2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.service.mds;
+
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.proto.DatasetModuleMeta;
+import co.cask.cdap.proto.DatasetTypeMeta;
+import co.cask.cdap.proto.id.DatasetModuleId;
+import co.cask.cdap.proto.id.DatasetTypeId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.StructuredTableContext;
+import co.cask.cdap.spi.data.TableNotFoundException;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.Fields;
+import co.cask.cdap.spi.data.table.field.Range;
+import co.cask.cdap.store.StoreDefinition;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Ranges;
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset types and modules metadata store
+ */
+public class DatasetTypeTable {
+
+  private static final Gson GSON = new Gson();
+
+  private final StructuredTable typeTable;
+  private final StructuredTable moduleTable;
+
+  public DatasetTypeTable(StructuredTable moduleTable, StructuredTable typeTable) {
+    this.moduleTable = moduleTable;
+    this.typeTable = typeTable;
+  }
+
+  public static DatasetTypeTable create(StructuredTableContext context) throws TableNotFoundException {
+    return new DatasetTypeTable(context.getTable(StoreDefinition.DatasetTypeStore.MODULE_TYPES),
```

Comment:
similar to what we did in AppMetaStore, can you instantiate the tables only when you access them?